# Chapter 7

# Loops 

### 7.1 Loops with while()

In [1]:
x = 100
while x/3 > 1:
    print(x/3) 
    x = x/3

33.333333333333336
11.111111111111112
3.703703703703704
1.234567901234568


Vòng lặp này lấy giá trị khởi tạo x, liên tục chia cho 3 và in ra kết quả, cho tới khi kết quả chia nhỏ hơn hoặc bằng 1. Nó minh họa quá trình lặp và giảm dần theo tỷ lệ (x giảm theo lũy thừa 1/3 mỗi lần).

In [2]:
print(x/3>1)

False


Dòng này kiểm tra điều kiện x/3 > 1 và in ra kết quả đúng/sai.

In [3]:
print(x/3)

0.41152263374485604


Dòng này chỉ đơn giản tính x chia 3 và in ra kết quả.

### 7.2 Loops with for()

In [5]:
print(list(range(0, 4)))

[0, 1, 2, 3]


Tạo một danh sách các số nguyên từ 0 đến 3 và in kết quả ra màn hình.

In [6]:
print(list(range(4)))

[0, 1, 2, 3]


Tạo một danh sách các số nguyên từ 0 đến 3 và in ra, giống như list(range(0, 4)), nhưng ngắn gọn hơn.

In [7]:
print(list(range(2, 10))) 

[2, 3, 4, 5, 6, 7, 8, 9]


Tạo và in danh sách các số nguyên từ 2 đến 9.

In [8]:
print(list(range(2, 10, 3)))

[2, 5, 8]


Tạo và in danh sách các số nguyên từ 2 đến 9, mỗi số cách số trước 3 đơn vị.

In [9]:
message = "The squared value of {} is {}" 
n=10
for i in range(0, n+1):
    print(message.format(i,i**2))

The squared value of 0 is 0
The squared value of 1 is 1
The squared value of 2 is 4
The squared value of 3 is 9
The squared value of 4 is 16
The squared value of 5 is 25
The squared value of 6 is 36
The squared value of 7 is 49
The squared value of 8 is 64
The squared value of 9 is 81
The squared value of 10 is 100


Vòng lặp này tính bình phương của các số từ 0 đến n và in kết quả theo định dạng chuỗi.

In [10]:
n=10
n_squares = []
for i in range(0, n+1):
    n_squares.append(i**2) 
    
print(n_squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Vòng lặp này tạo danh sách tất cả các bình phương từ 0 đến n.

In [11]:
message = "The squared value of {} is {}" 
for i in [0, 1, 2, 8, 9, 10]:
    print(message.format(i,i**2))

The squared value of 0 is 0
The squared value of 1 is 1
The squared value of 2 is 4
The squared value of 8 is 64
The squared value of 9 is 81
The squared value of 10 is 100


Vòng lặp này tính bình phương của các số cụ thể trong danh sách và in ra kết quả.

In [12]:
message = "There is(are) {} letter(s) in the name: {}"
for first_name in ["Pascaline", "Gauthier", "Xuan", "Jimmy"]:
  print(message.format(len(first_name), first_name))

There is(are) 9 letter(s) in the name: Pascaline
There is(are) 8 letter(s) in the name: Gauthier
There is(are) 4 letter(s) in the name: Xuan
There is(are) 5 letter(s) in the name: Jimmy


Vòng lặp này duyệt qua danh sách tên, tính số ký tự của từng tên và in ra theo định dạng đẹp mắt.

In [1]:
message = "i equals {} and j equals {}" 
for i in range(0,3):
    for j in range(0,3): 
        print(message.format(i, j))

i equals 0 and j equals 0
i equals 0 and j equals 1
i equals 0 and j equals 2
i equals 1 and j equals 0
i equals 1 and j equals 1
i equals 1 and j equals 2
i equals 2 and j equals 0
i equals 2 and j equals 1
i equals 2 and j equals 2


Là vòng lặp lồng nhau, in tất cả cặp giá trị (i, j) từ 0 đến 2. Đây là cách cơ bản để duyệt qua một ma trận hoặc tất cả kết hợp hai tập hợp số.

In [2]:
message = "New value for j: {}" 
j = 10
for i in range(0, 4):
    j += 5 
    print(message.format(j))

New value for j: 15
New value for j: 20
New value for j: 25
New value for j: 30


Vòng lặp này tăng giá trị j lên 5 mỗi lần và in kết quả, tổng cộng 4 lần. Đây là ví dụ cơ bản về vòng lặp với thay đổi giá trị biến theo từng bước.

In [3]:
print(j)

30


print(j) cho bạn biết giá trị cuối cùng của biến j sau khi vòng lặp hoặc các phép toán thay đổi nó.

### 7.3 Exercises

##### 1. Write a very naive program to determine if a number is prime or not. To do this:

1. define a number variable containing a natural integer of your choice (not too large),

In [4]:
# 1. Định nghĩa một biến number chứa một số nguyên tự nhiên
number = 19  

# 2. Kiểm tra xem số đó có phải là số nguyên tố hay không
is_prime = True  

# Một số nguyên tố phải lớn hơn 1
if number < 2:
    is_prime = False
else:
    # Kiểm tra chia hết từ 2 đến number-1
    for i in range(2, number):
        if number % i == 0:
            is_prime = False
            break

# 3. In kết quả
if is_prime:
    print(f"{number} is a prime number.")
else:
    print(f"{number} is not a prime number.")

19 is a prime number.


2. using a loop, check if each integer up to the square root of your number, is a divisor of your number (stop if ever it is the case)

In [6]:
import math  

# 1. Định nghĩa số cần kiểm tra
number = 19  

# 2. Giả sử ban đầu số là số nguyên tố
is_prime = True

# Loại các số < 2
if number < 2:
    is_prime = False
else:
    # Kiểm tra các số từ 2 đến căn bậc hai của number
    for i in range(2, int(math.sqrt(number)) + 1):
        if number % i == 0:
            is_prime = False
            break  

# 3. In kết quả
if is_prime:
    print(f"{number} is a prime number.")
else:
    print(f"{number} is not a prime number.")

19 is a prime number.


3. at the loop output, write a conditional instruction indicating whether or not the number is a prime one.

In [7]:
import math

# 1. Định nghĩa số cần kiểm tra
number = 17  

# 2. Giả sử ban đầu số là số nguyên tố
is_prime = True

# Loại các số < 2
if number < 2:
    is_prime = False
else:
    # Kiểm tra các số từ 2 đến căn bậc hai của number
    for i in range(2, int(math.sqrt(number)) + 1):
        if number % i == 0:
            is_prime = False
            break  

# 3. In ra kết quả bằng điều kiện
if is_prime:
    print(f"{number} is a prime number.")
else:
    print(f"{number} is not a prime number.")

17 is a prime number.


2. Choose a ‘mystery’ number between 1 and 100, and store it in an object called mystery_number. Then, create a loop that at each iteration performs a random draw of an integer between 1 and 100. As long as the number drawn is different from the mystery number, the loop must continue. At the output of the loop, a variable called nb_drawings will contain the number of draws made to obtain the mystery number.
Note: to draw a random number between 1 and 100, the method randint() of the module random may help).

In [8]:
import random  # để dùng hàm randint

# 1. Chọn một số bí ẩn từ 1 đến 100
mystery_number = random.randint(1, 100)
print(f"Mystery number (for verification): {mystery_number}")  

# 2. Khởi tạo biến đếm số lần rút thăm
nb_drawings = 0

# 3. Lặp cho đến khi rút được số bí ẩn
while True:
    draw = random.randint(1, 100)  # rút một số ngẫu nhiên từ 1 đến 100
    nb_drawings += 1  # tăng số lần rút thăm
    if draw == mystery_number:
        break  # dừng vòng lặp khi rút trúng số bí ẩn

# 4. In kết quả
print(f"The mystery number {mystery_number} was found after {nb_drawings} draws.")

Mystery number (for verification): 24
The mystery number 24 was found after 134 draws.


3. Use a loop to scan integers from 1 to 20 using a for loop, displaying in the console at each iteration if the current number is even.

In [9]:
for i in range(1, 21):
    if i % 2 == 0:
        print(f"{i} is even")

2 is even
4 is even
6 is even
8 is even
10 is even
12 is even
14 is even
16 is even
18 is even
20 is even


4. Use a for() loop to repeat the Fibonacci sequence until its tenth term (the Fn sequence is defined by the following recurrence relationship: Fn = Fn−1 + Fn−2; the initial values are F0 = 0 and F1 = 1).

In [1]:
# 1. Khởi tạo hai số đầu tiên của dãy Fibonacci
F0 = 0
F1 = 1

# 2. Tạo danh sách lưu trữ dãy Fibonacci
fib_sequence = [F0, F1]

# 3. Vòng lặp tính từ số thứ 2 đến số thứ 9 (tổng cộng 10 số)
for n in range(2, 10):
    Fn = fib_sequence[n - 1] + fib_sequence[n - 2]  # Fn = Fn-1 + Fn-2
    fib_sequence.append(Fn)

# 4. In kết quả
print("The first 10 terms of the Fibonacci sequence are:")
print(fib_sequence)

The first 10 terms of the Fibonacci sequence are:
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


# Kết thúc Chapter 7

# Chapter 8

# Functions

### 8.1 Definition

In [3]:
def square(x): 
    return x**2

In [4]:
print(square(2))

4


Hàm square(x) tách phần tính toán bình phương thành một khối riêng, giúp code dễ đọc, dễ tái sử dụng. return sẽ cho phép lấy kết quả của hàm để dùng ở các phần khác trong chương trình. Gọi hàm bằng square(2) sẽ thực hiện tính toán với tham số cụ thể.

In [5]:
print(square(-3))

9


Hàm square(x) tính bình phương của số x và trả kết quả in ra màn hình.

#### 8.1.1 Adding a Description

In [7]:
def square(x):
    """returns the squared value of x""" 
    return x**2

Hàm này tính bình phương của một số và trả kết quả, docstring giúp giải thích chức năng hàm.

#### 8.1.2 Arguments of a Function

In [10]:
def production(l, k, m): 
    """
    Returns the value of the production according to
    labour, capital and materials
    
    Keyword arguments:
    l -- labour (float)
    k -- capital (float)
    m -- materials (float)
    """
    return l**0.3 * k**0.5 * m**(0.2)

Hàm này tính giá trị sản xuất dựa trên lao động, vốn và nguyên liệu và trả kết quả, docstring giúp mô tả rõ ràng mục đích và tham số của hàm.

##### 8.1.2.1 Call Without Argument Names

In [11]:
prod_val = production(60, 42, 40) 
print(prod_val)

46.289449781254994


Đây là cách gọi hàm với các tham số cụ thể, lưu kết quả vào biến và in ra, giúp tái sử dụng kết quả cho các tính toán tiếp theo.

##### 8.1.2.2 Positional Arguments, Arguments by Keywords

In [12]:
def production_2(l, m, k=42): 
    """
    Returns the value of the production according to
    labour, capital and materials
    
    Keyword arguments:
    l -- labour (float)
    m -- materials (float)
    k -- capital (float) (default 42) 
    """
    return l**0.3 * k**0.5 * m**(0.2)

Hàm này tính giá trị sản xuất dựa trên lao động, nguyên liệu và vốn, trong đó tham số vốn có thể bỏ qua để dùng giá trị mặc định, giúp hàm linh hoạt hơn.

In [14]:
# By naming all argument, by ommitting k
prod_val_1 = production_2(l = 42, m = 40)
# By naming all argument and specifying k 
prod_val_2 = production_2(l = 42, m = 40, k = 42) 
# By naming only the argument k
prod_val_3 = production_2(42, 40, k = 42)
# Without naming any argument
prod_val_4 = production_2(42, 40, 42)
res = [prod_val_1, prod_val_2, prod_val_3, prod_val_4] 
print(res)

[41.59215573604822, 41.59215573604822, 41.59215573604822, 41.59215573604822]


Đây là cách gọi hàm linh hoạt trong Python bằng tên, bằng vị trí, kết hợp cả hai, và sử dụng tham số mặc định.

In [1]:
def production_3(a, l, m = 40, k=42): 
    """
    Returns the value of the production according to
    labour, capital and materials
    
    Keyword arguments:
    a -- total factor productivity (float) l -- labour (float)
    m -- materials (float) (default 40)
    k -- capital (float) (default 42)
    """
    return a * l**0.3 * k**0.5 * m**(0.2)

prod_val_1 = production_3(1, 42, m = 38) 
prod_val_2 = production_3(a = 1, l = 42) 
prod_val_3 = production_3(l = 42, a = 1) 
prod_val_4 = production_3(m = 40, l = 42, a = 1)

res = [prod_val_1, prod_val_2, prod_val_3, prod_val_4] 
print(res)

[41.16765711449734, 41.59215573604822, 41.59215573604822, 41.59215573604822]


Đây là cách gọi hàm linh hoạt trong Python bằng tên, bằng vị trí, kết hợp cả hai, và sử dụng tham số mặc định.

##### 8.1.2.3 Function as an Argument to Another Function

In [2]:
def square(x):
    """Returns the squared value of x""" 
    return x**2
    
def apply_fun_to_4(fun):
    """Applies the function `fun` to 4""" 
    return fun(4)
    
print(apply_fun_to_4(square))

16


Minh họa cách truyền một hàm như đối số cho hàm khác. Hàm square tính bình phương một số, còn hàm apply_fun_to_4 nhận một hàm khác và áp dụng nó cho số 4. Khi gọi apply_fun_to_4(square), Python thực hiện square(4) và trả kết quả 16. Ý nghĩa chung là có thể viết các hàm linh hoạt, tái sử dụng cao, áp dụng cho nhiều hàm khác nhau mà không phải viết lại logic.

### 8.2 Scope of a Function

In [3]:
# Definition of a global variable:
value = 1
# Definition of a local variable in function f
def f(x):
    value = 2
    new_value = 3
    print("value equals: ", value) 
    print("new_value equals: ", new_value) 
    return x + value

In [4]:
res = f(3)

value equals:  2
new_value equals:  3


Các biến cục bộ trong hàm (value = 2, new_value = 3) chỉ tồn tại bên trong hàm và dùng để tính toán hoặc in ra trong hàm, trong khi biến toàn cục (value = 1) vẫn giữ nguyên ngoài hàm. Khi gọi hàm, Python dùng giá trị của biến cục bộ để thực hiện phép tính và trả kết quả (res = x + value), nhưng không thay đổi biến toàn cục.

In [5]:
print("value equals: ", value)

value equals:  1


In ra giá trị hiện tại của biến value tại vị trí gọi lệnh. Nếu gọi ngoài hàm, Python sẽ in giá trị biến toàn cục (value = 1). Nếu gọi trong hàm và có biến cục bộ value, Python sẽ in giá trị biến cục bộ (value = 2).

In [6]:
print("new_value equals: ", new_value)

NameError: name 'new_value' is not defined

Các biến cục bộ chỉ dùng được trong hàm, còn ngoài hàm phải dùng biến toàn cục hoặc trả giá trị từ hàm ra nếu muốn sử dụng.

In [7]:
value = 1 
def f(x):
    return x + value 
    
print(f(2))

3


Trong hàm f(x), không có biến cục bộ value, nên Python sử dụng giá trị biến toàn cục. return x + value → 2 + 1 = 3 nên print(f(2)) in ra kết quả 3.

In [9]:
# The value variable is not defined in 
# the local environment of g().
# Python will then search in f(). 
value = 1
def f(): 
    value = 2 
    def g(x):
        return x + value 
    
    return g(2)
print(f())

4


Trong f(), value = 2 → biến cục bộ của hàm f. g(x) không định nghĩa biến value riêng → Python sẽ tìm biến trong môi trường bao quanh (enclosing scope), tức là biến cục bộ value của f(). g(2) → tính 2 + value = 2 + 2 = 4 nên f() trả về g(2) = 4 và print(f()) in ra 4.

In [13]:
# The value variable is not defined in g() or f() 
# but in the higher environment (here, global) 
value = 1
def f():
    def g(x):
        return x + value
    return g(2)
print(f())

3


Khi biến không tồn tại trong hàm hay hàm bao quanh, Python sẽ tìm biến trong phạm vi toàn cục.

In [14]:
def f(x): 
    global y 
    y = x+1

f(3)
print(y)

4


Từ khóa global cho phép tạo hoặc sửa biến toàn cục bên trong hàm. Không dùng global thì biến trong hàm sẽ là cục bộ, không ảnh hưởng biến ngoài hàm.

### 8.3 Lambda Functions

In [15]:
def square(x): 
    return x**2

In [16]:
square_2 = lambda x: x**2 
print(square_2(4))

16


lambda cho phép tạo hàm ngắn gọn trong một dòng mà không cần dùng def. gán hàm này cho biến square_2, để có thể gọi như hàm bình thường. square_2(4) → tính 4**2 = 16.

In [17]:
def production(l, k, m): 
    """
    Returns the value of the production according to
    labour, capital and materials.
    
    Keyword arguments:
    l -- labour (float)
    k -- capital (float)
    m -- materials (float)
    """
    return l**0.3 * k**0.5 * m**(0.2)

In [18]:
production_2 = lambda l,k,m : l**0.3 * k**0.5 * m**(0.2) 
print(production(42, 40, 42))

40.987803063838406


Python cho phép định nghĩa hàm bằng def hoặc lambda. Cả hai cách đều có thể tính toán và trả kết quả giống nhau, nhưng def thuận tiện khi muốn viết docstring hoặc thân hàm phức tạp, còn lambda tiện cho hàm ngắn gọn, dùng một lần.

In [19]:
print(production_2(42, 40, 42))

40.987803063838406


production_2 là hàm lambda tính giá trị sản xuất theo công thức Cobb-Douglas production = 𝑙^0.3 x 𝑘^0.5 x 𝑚^0.2. Gọi production_2(42, 40, 42) → tính giá trị với l=42, k=40, m=42.

### 8.4 Returning Several Values

In [20]:
import statistics 
def desc_stats(x):
    """Returns the mean and standard deviation of `x`"""
    return {"mean": statistics.mean(x), 
            "std_dev": statistics.stdev(x)}
    
x = [1,3,2,6,4,1,8,9,3,2] 
res = desc_stats(x) 
print(res)

{'mean': 3.9, 'std_dev': 2.8460498941515415}


import statistics → nhập module hỗ trợ các phép toán thống kê.

desc_stats(x) → hàm nhận một danh sách x, trả về dictionary gồm:

"mean": trung bình cộng của các giá trị.

"std_dev": độ lệch chuẩn (standard deviation).

Gọi hàm với danh sách x = [1,3,2,6,4,1,8,9,3,2] → tính mean = 3.9, std_dev ≈ 2.846 vàin ra kết quả.

In [21]:
message = "The average value equals {} and the standard deviation is {}" 
print(message.format(res["mean"], res["std_dev"]))

The average value equals 3.9 and the standard deviation is 2.8460498941515415


Sử dụng str.format() để định dạng chuỗi và in kết quả thống kê một cách dễ đọc.

### 8.5 Exercise

1. Create a function named sum_n_integers which returns the sum of the first integer n. Its only argument will be n.

In [22]:
def sum_n_integers(n):
    """
    Trả về tổng của n số nguyên đầu tiên
    """
    return n * (n + 1) // 2

print(sum_n_integers(5))  
print(sum_n_integers(10)) 

15
55


2. Using a loop, display the sum of the first 2 integers, then 3 first integers, then 4 first integers, etc. up to 10.

In [23]:
# Sử dụng vòng lặp để tính tổng dần từ 2 đến 10 số nguyên đầu tiên
for n in range(2, 11):  
    total = sum(range(1, n+1))  
    print(f"Sum of the first {n} integers is {total}")

Sum of the first 2 integers is 3
Sum of the first 3 integers is 6
Sum of the first 4 integers is 10
Sum of the first 5 integers is 15
Sum of the first 6 integers is 21
Sum of the first 7 integers is 28
Sum of the first 8 integers is 36
Sum of the first 9 integers is 45
Sum of the first 10 integers is 55


3. Create a function that from two points represented by pairs of coordinates (x1, y1) and (x2, y2) returns the Euclidean distance between these two points.
Propose a second solution using a lambda function.

In [25]:
import math

def euclidean_distance(p1, p2):
    """
    Trả về khoảng cách Euclid giữa hai điểm p1 và p2
    p1, p2: tuple chứa tọa độ (x, y)
    """
    x1, y1 = p1
    x2, y2 = p2
    return math.sqrt((x2 - x1)**2 + (y2 - y1)**2)

print(euclidean_distance((1, 2), (4, 6)))

5.0


# Kết thúc Chapter 8

# Chapter 9

#  Introduction to Numpy

In [11]:
 import numpy as np

### 9.1 Arrays

In [9]:
from random import random
from operator import truediv
l1 = [random() for i in range(1000)]
l2 = [random() for i in range(1000)]
# %timeit s = sum(map(truediv,l1,l2))

In [12]:
a1 = np.array(l1)
a2 = np.array(l2)
# %timeit s = np.sum(a1/a2)

Đoạn code tạo 2 dãy số ngẫu nhiên rồi so sánh tốc độ cộng kết quả chia từng phần tử bằng Python thuần và NumPy.

####  9.1.1 Creation

In [6]:
list = [1,2,4]
table = np.array(list)
print(table)

[1 2 4]


Chuyển list [1, 2, 4] thành mảng NumPy và in ra [1 2 4].

In [7]:
print(type(table))

<class 'numpy.ndarray'>


In ra kiểu dữ liệu của table là numpy.ndarray.

In [15]:
 list_2 = [ [1,2,3], [4,5,6] ]
 table_2 = np.array(list_2)
 print(table_2)

[[1 2 3]
 [4 5 6]]


Tạo mảng NumPy 2 chiều từ list [[1, 2, 3], [4, 5, 6]] và in ra

In [9]:
print(type(table_2))

<class 'numpy.ndarray'>


In ra kiểu dữ liệu của table_2 là numpy.ndarray.

In [11]:
tup = (1, 2, 3)
table = np.array(tup)
print(table)

[1 2 3]


Chuyển tuple (1, 2, 3) thành mảng NumPy và in ra \[1 2 3].

In [12]:
 print(type(table))

<class 'numpy.ndarray'>


In ra kiểu dữ liệu của table là numpy.ndarray.

In [13]:
 table = np.array([3, 2, 5, 1, 6, 5])
 table.shape = (3,2)
 print(table)

[[3 2]
 [5 1]
 [6 5]]


Tạo mảng NumPy từ [3, 2, 5, 1, 6, 5], đổi thành ma trận 3 hàng 2 cột và in ra

In [14]:
 print( np.zeros(4) )

[0. 0. 0. 0.]


Tạo mảng gồm 4 phần tử 0.0 và in ra \[0. 0. 0. 0.].

In [16]:
print( np.zeros(4, dtype = "int") )

[0 0 0 0]


Tạo mảng gồm 4 phần tử 0 kiểu số nguyên và in ra [0 0 0 0].

In [18]:
x = np.zeros(4, dtype = "int")
print(x, x.dtype)

[0 0 0 0] int64


Tạo mảng [0 0 0 0] kiểu số nguyên và in ra [0 0 0 0] int64.

In [19]:
y = x.astype("float")
print(x, x.dtype)

[0 0 0 0] int64


Tạo mảng [0 0 0 0] kiểu số nguyên và in ra [0 0 0 0] int64.

In [20]:
print(y, y.dtype)

[0. 0. 0. 0.] float64


Tạo mảng \[0. 0. 0. 0.] kiểu số thực và in ra \[0. 0. 0. 0.] float64.

Tạo mảng 2 hàng 3 cột toàn số 0.0 và in ra

In [21]:
print( np.zeros((2, 3, 4)) )

[[[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]

 [[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]]


Tạo mảng 3 chiều kích thước 2×3×4 toàn số 0.0 và in ra các khối giá trị 0.0.

In [31]:
print( np.empty((2, 3), dtype = "int") )

[[9218868437227405312 9218868437227405312 9218868437227405312]
 [9218868437227405312 9218868437227405312 9218868437227405312]]


Tạo mảng 2 hàng 3 cột kiểu số nguyên nhưng giá trị ngẫu nhiên (không khởi tạo) và in ra các số rác có sẵn trong bộ nhớ.

In [24]:
 print( np.ones((2, 3), dtype = "float") )

[[1. 1. 1.]
 [1. 1. 1.]]


Tạo mảng 2 hàng 3 cột toàn số 1.0 và in ra

In [25]:
 print( np.full((2, 3), 10, dtype = "float") )

[[10. 10. 10.]
 [10. 10. 10.]]


Tạo mảng 2 hàng 3 cột toàn số 10.0 và in ra

In [26]:
 print( np.full((2, 3), np.inf) )

[[inf inf inf]
 [inf inf inf]]


Tạo mảng 2 hàng 3 cột toàn giá trị vô cực và in ra

In [32]:
 print( np.eye(2, dtype="int64") )

[[1 0]
 [0 1]]


Tạo ma trận đơn vị 2×2 kiểu số nguyên 64-bit và in ra

In [33]:
 print( np.eye(3, k=-1) )

[[0. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]]


Tạo ma trận đơn vị 3×3 nhưng đường chéo chính được dịch xuống 1 hàng (k = -1) và in ra

In [34]:
print( np.identity(3, dtype = "int") )

[[1 0 0]
 [0 1 0]
 [0 0 1]]


Tạo ma trận đơn vị 3×3 kiểu số nguyên và in ra

In [13]:
np.arange( start, stop, step, dtype )

NameError: name 'start' is not defined

In [3]:
print( np.arange(5) )

[0 1 2 3 4]


In [4]:
print( np.arange(2, 5) )

[2 3 4]


In [5]:
print( np.arange(2, 10, 2) )

[2 4 6 8]


np.arange() tạo mảng số cách đều theo giá trị start, stop, step.

####  9.1.2 Dimensions

In [10]:
 print("ndim tableau : ", table.ndim)

ndim tableau :  2


In [11]:
print("ndim table_2 : ", table_2.ndim)

ndim table_2 :  2


In [12]:
 print("size table : ", table.size)

size table :  6


In [13]:
print("size table_2: ", table_2.size)

size table_2:  12


In [18]:
print("np.size(table):", np.size(table))

np.size(table): 6


In [15]:
print("size table: ", table.shape)

size table:  (2, 3)


In [19]:
 print("size table_2: ", table_2.shape)

size table_2:  (2, 3)


In ra số chiều của mảng table.

####  9.1.3 Extracting Elements from an Array

In [7]:
table_1 = np.arange(1,13)
table_2 = [ [1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]]
table_2 = np.array(table_2)

In [10]:
 message = "table_{}[0] : {} (type : {})"
 print(message.format(0, table_1[0], type(table_1[0])))

table_0[0] : 1 (type : <class 'numpy.int64'>)


In một dòng: Hiển thị giá trị đầu tiên của table\_1 cùng chỉ số và kiểu dữ liệu của nó.

In [24]:
 print(message.format(1, table_2[0], type(table_2[0])))

table_1[0] : [1 2 3] (type : <class 'numpy.ndarray'>)


Hiển thị giá trị đầu tiên của table\_2, kèm chỉ số 1 và kiểu dữ liệu của giá trị đó.


In [25]:
print("table_1[-1] : ", table_1[-1]) # last element

table_1[-1] :  12


Hiển thị phần tử cuối cùng của table\_1.


In [26]:
 print("table_2[-1] : ", table_2[-1]) # last element

table_2[-1] :  [10 11 12]


Hiển thị phần tử cuối cùng của table\_2.

In [28]:
# the elements from the 2nd (not included) to the 4th
print("Slice Table 1 : \n", table_1[2:4])


Slice Table 1 : 
 [3 4]


Hiển thị các phần tử từ vị trí thứ 3 đến thứ 4 của table\_1.


In [29]:
print("Sclie Table 2 : \n", table_2[2:4])

Sclie Table 2 : 
 [[ 7  8  9]
 [10 11 12]]


Hiển thị các phần tử từ vị trí thứ 3 đến thứ 4 của table\_2.


In [30]:
 # Within the 3rd element, access the 1st element
 print(table_2[2][0])

7


Truy cập phần tử đầu tiên bên trong phần tử thứ 3 của table\_2.

In [31]:
 print(table_2[2,0])

7


Tiếp tục thao tác với phần tử, cắt lát, kiểm tra kiểu hoặc lặp qua danh sách/mảng.

In [33]:
print("Second column: \n", table_2[:, [1]])

Second column: 
 [[ 2]
 [ 5]
 [ 8]
 [11]]


Lấy cột thứ hai của table\_2 (mọi hàng, chỉ cột 1) nếu table\_2 là numpy array.


In [34]:
 print("Second and third columns: \n", table_2[:, [1,2]])

Second and third columns: 
 [[ 2  3]
 [ 5  6]
 [ 8  9]
 [11 12]]


Lấy cột thứ hai và thứ ba của table\_2 (mọi hàng, chỉ cột 1 và 2) nếu table\_2 là numpy array.

In [35]:
 print("2nd and 4th elements: \n", table_2[[1,3]])

2nd and 4th elements: 
 [[ 4  5  6]
 [10 11 12]]


Lấy phần tử thứ 2 và thứ 4 của table\_2 nếu table\_2 là numpy array.

#####  9.1.3.1 Extraction Using Boolean

In [36]:
 table = np.array([0, 3, 2, 5, 1, 4])
 res = table[[True, False, True, False, True, True]]
 print(res)

[0 2 1 4]


Chọn các phần tử của table tương ứng với True trong mảng boolean.

In [37]:
 mask = table % 2 == 0
 print(mask)

[ True False  True False False  True]


Tạo mảng boolean đánh dấu các phần tử chẵn trong table.


In [38]:
 print(type(mask))

<class 'numpy.ndarray'>


Hiển thị kiểu dữ liệu của mask, là numpy ndarray.


In [39]:
print(table[mask])

[0 2 4]


Chọn các phần tử chẵn trong table dựa trên mask.


#### 9.1.4 Modification

In [40]:
table = np.array([ [1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12]])
table[0] = [11, 22, 33]
print(table)

[[11 22 33]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


Thay hàng đầu tiên bằng \[11, 22, 33] và in mảng.


In [42]:
table[0] = 100
print(table)

[[100 100 100]
 [  4   5   6]
 [  7   8   9]
 [ 10  11  12]]


Đặt toàn bộ hàng đầu tiên thành 100 và in mảng.


In [43]:
table[0:2] = 100
print(table)

[[100 100 100]
 [100 100 100]
 [  7   8   9]
 [ 10  11  12]]


Đặt toàn bộ hai hàng đầu tiên thành 100 và in mảng.


In [44]:
table[:] = 0
print(table)

[[0 0 0]
 [0 0 0]
 [0 0 0]
 [0 0 0]]


Đặt tất cả phần tử của table thành 0 và in mảng.


####  9.1.4.1 Insterting Elements

In [45]:
t_1 = np.array([1,3,5])
print("t_1 : ", t_1)

t_1 :  [1 3 5]


Tạo mảng t\_1 với các phần tử \[1, 3, 5] và in ra.


In [17]:
 t_2 = np.array([[1,2,3], [5,6,7]])
 print("t_2 : \n", t_2)

t_2 : 
 [[1 2 3]
 [5 6 7]]


In [13]:
t_1 = np.append(t_1, 1)
print("t_1 after the insertion: ", t_1)

t_1 after the insertion:  [1 3 5 1]


Thêm 1 vào cuối t\_1 và in mảng mới.


In [14]:
 t_2 = np.array([[1,2,3], [5,6,7]])
 print("t_2 : \n", t_2)

t_2 : 
 [[1 2 3]
 [5 6 7]]


Tạo mảng 2 chiều t\_2 và in ra.


In [18]:
add_col_t_2 = np.array([[4], [8]])
t_2 = np.append(t_2,add_col_t_2, axis = 1)
print("t_2 after the insertion: \n", t_2)

t_2 after the insertion: 
 [[1 2 3 4]
 [5 6 7 8]]


Thêm cột \[4, 8] vào t\_2 và in ra.

In [18]:
ajout_ligne_t_2 = np.array([10, 11, 12, 13])
t_2 = np.vstack([t_2,ajout_ligne_t_2])
print("t_2 après ajout ligne : \n", t_2)

t_2 après ajout ligne : 
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


Thêm hàng \[10, 11, 12, 13] vào t\_2 và in mảng mới.

#####  9.1.4.2 Deleting / Removing Elements

In [19]:
 print("t_1 : ", t_1)
 # Remove the last element

t_1 :  [1 3 5 1]


Hiển thị t\_1 trước khi xóa phần tử cuối.


In [20]:
np.delete(t_1, (-1))

array([1, 3, 5])

Xóa phần tử cuối của t\_1 và trả về mảng mới.


In [21]:
print("t_1 : ", t_1)
 # Remove the first and second elements:

t_1 :  [1 3 5 1]


Hiển thị t\_1 trước khi xóa phần tử đầu và thứ hai.

In [22]:
 t_1 = np.delete(t_1, ([0, 2]))
 print(t_1)

[3 1]


Xóa phần tử đầu tiên và phần tử thứ ba (chỉ số 0 và 2) của t\_1 và in mảng mới.


In [23]:
print("t_2 : ", t_2)
 # Remove the last column:

t_2 :  [[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


Hiển thị t\_2 trước khi xóa cột cuối.


In [25]:
np.delete(t_2, (0), axis=1)

array([[ 2,  3,  4],
       [ 6,  7,  8],
       [11, 12, 13]])

Xóa cột đầu tiên của t\_2 (axis=1) và trả về mảng mới.


Delete multiple columns:

In [26]:
print("t_2 : ", t_2)
 # Remove the first and third columns:

t_2 :  [[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


Hiển thị t\_2 trước khi xóa cột thứ nhất và thứ ba.


In [27]:
np.delete(t_2, ([0,2]), axis=1)

array([[ 2,  4],
       [ 6,  8],
       [11, 13]])

Xóa cột thứ nhất và thứ ba của t\_2 (axis=1) và trả về mảng mới.


In [28]:
print("t_2 : ", t_2)
 # Remove the first line:

t_2 :  [[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


Hiển thị t\_2 trước khi xóa hàng đầu tiên.

In [29]:
np.delete(t_2, (0), axis=0)

array([[ 5,  6,  7,  8],
       [10, 11, 12, 13]])

Xóa hàng đầu tiên của t\_2 (axis=0) và trả về mảng mới.


In [30]:
print("t_2 : ", t_2)
 # Remove the first and third lines:

t_2 :  [[ 1  2  3  4]
 [ 5  6  7  8]
 [10 11 12 13]]


Hiển thị t\_2 trước khi xóa hàng thứ nhất và thứ ba.


In [31]:
np.delete(t_2, ([0,2]), axis=0)

array([[5, 6, 7, 8]])

Xóa hàng thứ nhất và thứ ba của t\_2 (axis=0) và trả về mảng mới.

#### 9.1.5 Copyi of an Array

In [32]:
 table_1 = np.array([1, 2, 3])
 table_2 = table_1


In [33]:
 table_2[0] = 0
 print("Table 1: \n", table_1)

Table 1: 
 [0 2 3]


Đặt hàng đầu tiên của table\_2 bằng 0 nhưng in table\_1, không liên quan.


In [34]:
print("Table 2: \n", table_2)

Table 2: 
 [0 2 3]


Hiển thị table\_2 sau khi thay đổi hàng đầu tiên thành 0.


In [37]:
table_1 = np.array([1, 2, 3])
table_2 = np.array(table_1)
table_2[0] = 0
print("table_1 : ", table_1)

table_1 :  [1 2 3]


table\_1 không thay đổi, vẫn là \[1, 2, 3], vì table\_2 được sao chép riêng.


In [38]:
 print("table_2 : ", table_2)

table_2 :  [0 2 3]


Hiển thị table\_2, hàng đầu tiên đã bị thay đổi thành 0.


In [40]:
table_1 = np.array([1, 2, 3])
table_2 = table_1.copy()
table_2[0] = 0
print("table_1 : ", table_1)

table_1 :  [1 2 3]


table\_1 vẫn giữ nguyên \[1, 2, 3] vì table\_2 là bản sao riêng bằng copy().


In [41]:
 print("table_2 : ", table_2)

table_2 :  [0 2 3]


Hiển thị table\_2, phần tử đầu tiên đã đổi thành 0.

In [42]:
 table_1 = np.array([1, 2, 3, 4])
 table_2 = table_1[:2]
 table_2[0] = 0
 print("table_1 : ", table_1)

table_1 :  [0 2 3 4]


Phần tử đầu tiên của table\_1 cũng bị thay đổi thành 0 vì table\_2 là slice tham chiếu tới table\_1.


In [43]:
print("table_2 : ", table_2)

table_2 :  [0 2]


Hiển thị table\_2, phần tử đầu tiên đã đổi thành 0.


####  9.1.6 Sorting

In [44]:
 table = np.array([3, 2, 5, 1, 6, 5])
 print("Sorted Table: ", np.sort(table))

Sorted Table:  [1 2 3 5 5 6]


Sắp xếp các phần tử của table theo thứ tự tăng dần và in kết quả.


In [45]:
 print("Table: ", table)

Table:  [3 2 5 1 6 5]


Hiển thị table gốc, vẫn giữ nguyên thứ tự ban đầu.

In [19]:
 table = np.array([3, 2, 5, 1, 6, 5])
 table.sort()
 print("The array was modified: ", table)

The array was modified:  [1 2 3 5 5 6]


Sắp xếp table tại chỗ, thay đổi mảng gốc và in kết quả.


####  9.1.7 Transposition

In [20]:
 table = np.array([3, 2, 5, 1, 6, 5])
 table.shape = (3,2)
 print("Array: \n", table)

Array: 
 [[3 2]
 [5 1]
 [6 5]]


Chuyển table thành mảng 3 hàng 2 cột và in ra.

In [48]:
 print("Transposed Array: \n", table.T)

Transposed Array: 
 [[3 5 6]
 [2 1 5]]


In mảng table sau khi chuyển vị (hàng thành cột và cột thành hàng).

In [49]:
print(np.transpose(table))

[[3 5 6]
 [2 1 5]]


In mảng table sau khi chuyển vị, tương đương table.T.

In [52]:
table_transpose = np.transpose(table)
table_transpose[0,0] = 99
print("Array: \n", table)

Array: 
 [[99  2]
 [ 5  1]
 [ 6  5]]


Mảng gốc table cũng thay đổi phần tử \[0,0] thành 99 vì transpose trả về view, không phải bản sao.


In [53]:
print("Transpose of the Array: \n", table_transpose)

Transpose of the Array: 
 [[99  5  6]
 [ 2  1  5]]


Hiển thị mảng đã chuyển vị table\_transpose, phần tử \[0,0] là 99.


In [54]:
print("table: ", table.base)

table:  None


Hiển thị mảng gốc mà table đang tham chiếu; nếu None tức table không phải là view của mảng khác.


In [55]:
 print("table_transpose : ", table_transpose.base)

table_transpose :  [[99  2]
 [ 5  1]
 [ 6  5]]


Hiển thị mảng gốc mà table\_transpose đang tham chiếu; kết quả là table vì transpose trả về view.


#### 9.1.8 Operations on Arrays

##### 9.1.8.1 + and- Operators

In [21]:
t_1 = np.array([1, 2, 3, 4])
t_2 = np.array([5, 6, 7, 8])
t_3 = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
t_4 = np.array([[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]])
t_1 + t_2

array([ 6,  8, 10, 12])

Cộng từng phần tử tương ứng của t\_1 và t\_2, kết quả là \[6, 8, 10, 12].


In [57]:
 t_3 + t_4

array([[14, 16, 18, 20],
       [22, 24, 26, 28],
       [30, 32, 34, 36]])

Cộng từng phần tử tương ứng của t_3 và t_4, in kết quả 

In [58]:
t_1- t_2

array([-4, -4, -4, -4])

Trừ từng phần tử của t\_2 khỏi t\_1, kết quả là \[-4, -4, -4, -4].

In [59]:
 print("t_1 + 3 : \n", t_1 + 3)

t_1 + 3 : 
 [4 5 6 7]


Cộng 3 vào từng phần tử của t\_1 và in kết quả.


In [60]:
print("t_1 + 3. : \n", t_1 + 3.)

t_1 + 3. : 
 [4. 5. 6. 7.]


Cộng 3.0 (float) vào từng phần tử của t\_1, kết quả là mảng float.


In [61]:
 print("t_3 + 3 : \n", t_3 + 3)

t_3 + 3 : 
 [[ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]


Cộng 3 vào từng phần tử của t\_3 và in mảng kết quả.


In [62]:
 print("t_3- 3 : \n", t_3- 3)

t_3- 3 : 
 [[-2 -1  0  1]
 [ 2  3  4  5]
 [ 6  7  8  9]]


Trừ 3 từ từng phần tử của t\_3 và in mảng kết quả.


#####  9.1.8.2 * and / Operators

In [63]:
t_1 * t_2

array([ 5, 12, 21, 32])

Nhân từng phần tử tương ứng của t\_1 và t\_2, kết quả là \[5, 12, 21, 32].


In [64]:
 t_3 * t_4

array([[ 13,  28,  45,  64],
       [ 85, 108, 133, 160],
       [189, 220, 253, 288]])

Nhân từng phần tử tương ứng của t_3 và t_4, in kết quả

In [65]:
 t_3 / t_4

array([[0.07692308, 0.14285714, 0.2       , 0.25      ],
       [0.29411765, 0.33333333, 0.36842105, 0.4       ],
       [0.42857143, 0.45454545, 0.47826087, 0.5       ]])

Chia từng phần tử tương ứng của t_3 cho t_4, in kết quả

In [67]:
 print("t_1 * 3 : \n", t_1 * 3)

t_1 * 3 : 
 [ 3  6  9 12]


Nhân từng phần tử của t\_1 với 3 và in kết quả.

In [68]:
print("t_1 / 3 : \n", t_1 / 3)

t_1 / 3 : 
 [0.33333333 0.66666667 1.         1.33333333]


Chia từng phần tử của t\_1 cho 3 và in kết quả.


 ##### 9.1.8.3 Power

In [69]:
print("t_1 ** 3 : \n", t_1 ** 3)

t_1 ** 3 : 
 [ 1  8 27 64]


Lũy thừa từng phần tử của t\_1 lên mũ 3 và in kết quả.


##### 9.1.8.4 Operations on Matrices

In [70]:
 np.dot(t_3, t_4.T)

array([[150, 190, 230],
       [382, 486, 590],
       [614, 782, 950]])

Thực hiện nhân ma trận giữa t\_3 và chuyển vị của t\_4, kết quả là ma trận 3x3 với tích vô hướng của các hàng và cột tương ứng.


In [27]:
np.dot(t_3, t_4)

ValueError: shapes (3,4) and (3,4) not aligned: 4 (dim 1) != 3 (dim 0)

In [28]:
t_3 @ t_4.T

array([[150, 190, 230],
       [382, 486, 590],
       [614, 782, 950]])

Thực hiện nhân ma trận giữa t\_3 và chuyển vị của t\_4 bằng toán tử @, kết quả là ma trận 3x3.


In [29]:
np.dot(t_1, t_3.T)

array([ 30,  70, 110])

Nhân t\_1 (1x4) với t\_3.T (4x3), kết quả là mảng 1x3 chứa các tích vô hướng với từng hàng của t\_3.


####  9.1.9 Logical Operators

In [32]:
t = np.array([[1, 10, 3, 24], [9, 12, 40, 2], [0, 7, 2, 14]])
mask = np.logical_and(t <= 20, t >= 10)
print("mask: \n", mask)

mask: 
 [[False  True False False]
 [False  True False False]
 [False False False  True]]


Tạo mảng boolean mask đánh dấu các phần tử của t nằm trong khoảng từ 10 đến 20.


In [33]:
 print("the elements of t between 10 and 20: \n",
 t[mask])

the elements of t between 10 and 20: 
 [10 12 14]


Chọn và in các phần tử của t nằm trong khoảng từ 10 đến 20.


####  9.1.10 Some Constants

In [34]:
 print("Addition : ", np.nan + 1)

Addition :  nan


Kết quả là nan vì bất cứ phép tính nào với np.nan đều ra nan.


In [35]:
 print("Substraction : ", np.nan- 1)

Substraction :  nan


Kết quả là nan vì bất kỳ phép tính nào với np.nan đều không xác định.


In [36]:
 print("Multiplication : ", np.nan + 1)

Multiplication :  nan


Kết quả là nan vì mọi phép tính với np.nan đều không xác định.


In [37]:
 print("Division : ", np.nan / 1)

Division :  nan


Kết quả là nan vì bất kỳ phép tính nào với np.nan đều không xác định.


#### 9.1.11 Universal functions

In [38]:
 t_1 = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
 t_2 = np.array([[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]])
 np.log(t_1) # Natural Logarithm

array([[0.        , 0.69314718, 1.09861229, 1.38629436],
       [1.60943791, 1.79175947, 1.94591015, 2.07944154],
       [2.19722458, 2.30258509, 2.39789527, 2.48490665]])

Tính logarith tự nhiên (ln) của từng phần tử trong t\_1.


In [39]:
 np.subtract(t_1, t_2) # Substraction of the elements of `t_1` by those of `t_2`

array([[-12, -12, -12, -12],
       [-12, -12, -12, -12],
       [-12, -12, -12, -12]])

np.subtract(t\_1, t\_2) tra ve mot mang moi bang cach tru tung phan tu t\_2 cho tung phan tu t\_1


####  9.1.12 Mathematical and Statistical Methods and Functions

In [40]:
 t_1 = np.array([[1, 2, 3, 4], [-1, 6, 7, 8], [9,-1, 11, 12]])
 print("t_1 : \n", t_1)

t_1 : 
 [[ 1  2  3  4]
 [-1  6  7  8]
 [ 9 -1 11 12]]


t\_1 = np.array(\[\[1, 2, 3, 4], \[-1, 6, 7, 8], \[9, -1, 11, 12]]) in ra mang 2 chieu 3 hang 4 cot


In [41]:
 print("Sum of the elements: ", t_1.sum())

Sum of the elements:  61


print("Sum of the elements: ", t\_1.sum()) in ra tong tat ca phan tu trong mang t\_1


In [42]:
 print("Covariance of the elements: ", t_1.var())

Covariance of the elements:  18.07638888888889


print("Covariance of the elements: ", t\_1.var()) in ra phuong sai cua tat ca phan tu trong mang t\_1


In [44]:
 print("Sum per column: ", t_1.sum(axis=0))

Sum per column:  [ 9  7 21 24]


print("Sum per column: ", t\_1.sum(axis=0)) in ra tong cac phan tu theo tung cot trong mang t\_1


In [45]:
print("Sum per row: ", t_1.sum(axis=1))

Sum per row:  [10 20 31]


print("Sum per row: ", t\_1.sum(axis=1)) in ra tong cac phan tu theo tung hang trong mang t\_1


In [46]:
 t_1 = np.array([[1, 2, 3, 4], [-1, 6, 7, 8], [9,-1, 11, 12]])
 print("t_1 : \n", t_1)

t_1 : 
 [[ 1  2  3  4]
 [-1  6  7  8]
 [ 9 -1 11 12]]


t\_1 = np.array(\[\[1, 2, 3, 4], \[-1, 6, 7, 8], \[9, -1, 11, 12]]) tao mang 2 chieu 3 hang 4 cot va in ra no


In [47]:
print("Variance: ", np.var(t_1))

Variance:  18.07638888888889


print("Variance: ", np.var(t\_1)) in ra phuong sai cua tat ca phan tu trong mang t\_1


In [53]:
 t_1 = np.array([[1, 2, np.nan, 4], [-1, 6, 7, 8], [9,-1, 11, 12]])
 print("Sum: ", np.sum(t_1))

Sum:  nan


Code này tạo mảng t\_1 có giá trị np.nan và khi dùng np.sum(t\_1) se tra ve nan vi bat ky phep tinh tong nao co nan deu ra nan.


In [54]:
print("Sum ignoring NaN values: ", np.nansum(t_1))

Sum ignoring NaN values:  58.0


print("Sum ignoring NaN values: ", np.nansum(t\_1)) in ra tong cac phan tu trong mang t\_1 bo qua gia tri nan


In [55]:
 v_1 = np.array([1, 1, 4, 2])
 w = np.array([1, 1, .5, 1])
 print("Weighted average: ", np.average(v_1, weights=w))

Weighted average:  1.7142857142857142


print("Weightedaverage: ", np.average(v_1, weights=w)) tinh trung binh co trong so cua mang v_1 theo trong so w

### 9.2 Generation of Pseudo-random Numbers

In [56]:
from numpy import random

In [58]:
x = np.random.normal(size=10)
print(x)

[-0.71586133 -0.13090579  0.04743999 -0.69399062 -0.80627821 -0.48372051
  1.48785852  0.15374603 -3.87010636 -0.81932223]


x = np.random.normal(size=10) tao mot mang 10 phan tu ngau nhien tu phan phoi chuan va in ra no


In [59]:
 x = np.random.randn(10, 4)
 print(x)

[[ 9.41946925e-01  8.85578840e-01  6.96388161e-01 -8.51649959e-02]
 [-5.93557079e-01  4.88549896e-01 -9.79942728e-01 -1.92385133e-01]
 [ 7.52452027e-02 -1.02847030e+00  1.32356524e+00 -3.97279416e-01]
 [ 3.72981936e-01  4.84012490e-01 -1.74864731e+00  2.91496451e-01]
 [ 1.98097206e-01 -2.37604430e-01  1.76442247e-01  1.23185500e-01]
 [ 3.83246748e-01 -1.07011972e+00 -2.05703953e+00 -1.89883845e+00]
 [-5.75903401e-01 -2.55233063e-01 -4.51205722e-01 -5.10064430e-02]
 [-2.98977070e-01 -1.13804915e+00 -1.27337578e+00 -1.50223911e+00]
 [ 1.26533729e+00  1.86176673e-01 -2.24079003e+00  1.86382886e-01]
 [ 4.16376682e-01  4.31406250e-01  5.60046304e-04 -4.87036330e-01]]


x = np.random.randn(10, 4) tao mot mang 10 hang 4 cot gom cac so ngau nhien tu phan phoi chuan va in ra no


In [61]:
np.random.seed(1234)
x = np.random.normal(size=10)
print(x)

[ 0.47143516 -1.19097569  1.43270697 -0.3126519  -0.72058873  0.88716294
  0.85958841 -0.6365235   0.01569637 -2.24268495]


np.random.seed(1234) dat seed de tao so ngau nhien co the lap lai
x = np.random.normal(size=10) tao mang 10 phan tu ngau nhien tu phan phoi chuan va in ra no

In [63]:
 from numpy.random import RandomState
 rs = RandomState(123)
 x = rs.normal(10)
 print(x)

8.914369396699438


rs = RandomState(123) tao mot doi tuong RandomState co seed 123
x = rs.normal(10) sinh mot so ngau nhien tu phan phoi chuan (mean=0, std=1) va in ra no

In [64]:
 x = np.arange(10)
 y = np.random.permutation(x)
 print("x : ", x)

x :  [0 1 2 3 4 5 6 7 8 9]


x = np.arange(10) tao mang [0 1 2 3 4 5 6 7 8 9]
y = np.random.permutation(x) tao mot ban tron cua x khong thay doi x ban dau
print("x : ", x) in ra mang x ban dau

In [65]:
 print("y : ", y)

y :  [9 7 4 3 8 2 6 1 0 5]


print("y : ", y) in ra mang y da duoc tron ngau nhien tu x


In [66]:
 x = np.arange(10)
 print("x avant permutation : ", x)

x avant permutation :  [0 1 2 3 4 5 6 7 8 9]


x = np.arange(10) tao mang \[0 1 2 3 4 5 6 7 8 9] va print("x avant permutation : ", x) in ra mang x truoc khi tron ngau nhien


In [67]:
np.random.permutation(x)

array([7, 5, 4, 1, 0, 8, 3, 9, 6, 2])

np.random.permutation(x) tao mot ban tron ngau nhien cua mang x khong thay doi x ban dau


In [68]:
print("xaprès permutation: ",x)

xaprès permutation:  [0 1 2 3 4 5 6 7 8 9]


print("xaprès permutation: ", x) in ra mang x ban dau, van giu nguyen vi np.random.permutation() khong thay doi mang goc


### 9.3 Exercise

Firstexercise

In [1]:
import numpy as np

x = np.array([1, 2, 3, 4, 5])
print("Vector x: ", x)

Vector x:  [1 2 3 4 5]


#####  1. Create this vector using an array called x.

In [2]:
import numpy as np

x = np.array([1, 2, 3, 4, 5])

##### 2. Display the type of x and its length.

In [3]:
   print(type(x))  # <class 'numpy.ndarray'>
   print(len(x))   # 5

<class 'numpy.ndarray'>
5


##### 3. Extract the first element, then do the same with the last one.

In [4]:
   first_element = x[0]  # 1
   last_element = x[-1]   # 5

##### 4. Extract the first three elements and store them in a vector called a.

In [5]:
x = np.array([1, 2, 3, 4, 5])

##### 5. Extract the 1st, 2nd, and 5th elements of the vector; store them in a vector called b.

In [6]:
b = np.array([10, 20, 30, 40, 50])

##### 6. Add the number 10 to the vector x, then multiply the result by 2.

In [7]:
a = np.array([10, 20, 30, 40, 50])

##### 7. Add a and b, comment on the result.  

In [8]:
result_ab = a + b

##### 8. Make the following addition: x + a; comment on the result, then look at the result of a + x.

In [9]:
result_x_a = x + a
result_a_x = a + x

##### 9. Multiply the vector by the scalar c which will be set to 2

In [10]:
c = 2
x_scaled = x * c

##### 10. Multiply a and b; comment on the result.

In [11]:
result_a_b = a * b

##### 11. Perform the following multiplication: x * a; comment on the results.

In [12]:
result_x_a_multiplication = x * a

##### 12. Retrieve the positions of the multiples of 2 and store them in a vector called ind, then store only the multiples of 2 of x in a vector called mult_2.

In [13]:
ind = np.where(x % 2 == 0)[0]
mult_2 = x[x % 2 == 0]

##### 13. Display the elements of x that are multiples of 3 and multiples of 2.

In [14]:
multiples_3_and_2 = x[(x % 3 == 0) & (x % 2 == 0)]  

##### 14. Display the elements of x that are multiples of 3 or multiples of 2.

In [15]:
multiples_3_or_2 = x[(x % 3 == 0) | (x % 2 == 0)] 

##### 15. Calculate the sum of the elements of x.

In [16]:
sum_x = np.sum(x)

##### 16. Replace the first element of x with 4.

In [17]:
x[0] = 4

##### 17. Replace the first element of x with the value NaN, then calculate the sum of the elements of x.

In [18]:
x[0] = np.nan
sum_x_nan = np.nansum(x)

ValueError: cannot convert float NaN to integer

##### 18. Delete the vector x.

In [19]:
del x

Second exercise

##### 1. Create the following matrix:

In [20]:
   import numpy as np
   A = np.array([[-3, 5, 6], [-1, 2, 2], [1, -1, -1]])

##### 2. Display the size of A, its number of columns, its number of rows, and its length.

In [21]:
   print(A.shape)          # (3, 3) - size of A
   print(A.shape[0])       # 3 - number of rows
   print(A.shape[1])       # 3 - number of columns
   print(len(A))           # 3 - length (number of rows)

(3, 3)
3
3
3


##### 3. Extract the second column from A, then the first row.

In [22]:
   second_column = A[:, 1]  # array([ 5,  2, -1])
   first_row = A[0, :]       # array([-3,  5,  6])

##### 4. Extract the element in the third position in the first line.

In [8]:
element = A[0, 2]  # 6

##### 5. Extract the submatrix of dimension 2 × 2 from the lower corner of A.

In [9]:
submatrix = A[1:3, 1:3]  

##### 6. Calculate the sum of the columns and then the rows of A.

In [10]:
   column_sum = np.sum(A, axis=0)  
   row_sum = np.sum(A, axis=1)

##### 7. Display the diagonal of A.

In [11]:
diagonal = np.diag(A)  

##### 8. Add the vector to A and store it in an object called B.

In [12]:
vector = np.array([[1], [2], [3]])  # Column vector
B = np.hstack((A, vector))  # Add vector to the right of A

##### 9. Remove the fourth vector from B.

In [13]:
# Since B has only 4 columns, we can remove the last column (index 3)
B = B[:, :-1]  # Now B has only 3 columns

##### 10. Remove the first and third rows from B.

In [14]:
B = np.delete(B, [0, 2], axis=0)

##### 11. Add scalar 10 to A.

In [55]:
A_plus_10 = A + 10 

##### 12. Add the vector to the right of the matrix A and store the result in A.

In [56]:
A = np.hstack((A, vector))

##### 13. Add the identity matrix to A.

In [59]:
I = np.eye(3)  # Identity matrix of size 3
A = A + I  # Add identity matrix to A

##### 14. Divide all the elements of the matrix A by 2.

In [60]:
A = A / 2

##### 15. Multiply the matrix A by the line vector

In [16]:
line_vector = np.array([[1, 2, 3]])  # Row vector
result_multiplication = A @ line_vector.T  # Matrix multiplication

##### 16. Display the transposition of A.

In [17]:
A_transpose = A.T

##### 17. Perform the product with transposition 

In [18]:
product_transpose = A.T @ A

# Kết thúc Chapter 9